### PR TITLE
Migrate RCPurchasesErrorUtils

### LINKED ISSUE #548: Migrate PurchasesErrorUtils
~Blocked on #547~

### PR DESCRIPTION
**Pull request by vegaro**
Should fix #548

### CODE REVIEW

File: PurchasesCoreSwift/Public/Error Handling/BackendErrorCode.swift [hunk 1]
```diff
@@ -20,6 +20,7 @@ import Foundation
  */
 @objc(RCBackendErrorCode) public enum BackendErrorCode: Int, Error {
 
+    @objc(RCBackendUnknownError) case unknownError = 0
```
Comment (vegaro) [inline PurchasesCoreSwift/Public/Error Handling/BackendErrorCode.swift:23]: Added this one as a default case

File: PurchasesCoreSwift/Public/Error Handling/ErrorCode.swift [hunk 2]
```diff
@@ -46,3 +46,128 @@ import Foundation
...
+        @unknown default:
+            return "Something went wrong."
+        }
+
+    }
+
+    /**
+     * The error short string, based on the error code.
+     */
+    var codeName: String {
```
Comment (vegaro) [inline PurchasesCoreSwift/Public/Error Handling/ErrorCode.swift:116]: Open to suggestions for this naming

File: PurchasesCoreSwift/Public/Error Handling/ErrorUtils.swift [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+
+    }
+
+    static func backendUnderlyingError(backendCode: NSNumber?, backendMessage: String?) -> Error {
+
+        let error: Error
+        if let maybeBackendCode = backendCode,
+           // NOTE: Behavior change here, before we would make an error with the code that
+           // came from the backend. Now we create a BackendErrorCode or an Unknown
+           let backendError = BackendErrorCode.init(rawValue: maybeBackendCode.intValue) {
```
Comment (vegaro) [inline PurchasesCoreSwift/Public/Error Handling/ErrorUtils.swift:280]: Behavior change here, before we would make an NSError with the code that came in the JSON response from the backend. Now we create a BackendErrorCode or an BackendErrorCode.unknownError instead. Open to suggestions here.

File: PurchasesTests/Networking/BackendTests.swift [hunk 4]
```diff
@@ -1502,7 +1502,7 @@ class BackendTests: XCTestCase {
 
         let currentAppUserID = "old id"
         let underlyingErrorMessage = "header fields too large"
-        let underlyingErrorCode = "123456"
+        let underlyingErrorCode = BackendErrorCode.cannotTransferPurchase.rawValue // NOTE: changed because now we create unknown errors
```
Comment (vegaro) [inline PurchasesTests/Networking/BackendTests.swift:1505]: "123456" would be converted to `.unknownError` with the change from line https://github.com/RevenueCat/purchases-ios/pull/670/files#diff-c4fc22c0d43f6801ea744bd065af825839de0cbe05fdc979640850bcee2135afR278

File: PurchasesCoreSwift/Public/Error Handling/ErrorUtils.swift [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+    /**
+     * Constructs an NSError with the [RCNetworkError] code and a populated [RCUnderlyingErrorKey] in
+     * the [NSError.userInfo] dictionary.
+     *
+     * @param underlyingError The value of the [NSUnderlyingErrorKey] key.
+     *
+     * @note This error is used when there is an error performing network request returns an error or when there
+     * is an [NSJSONSerialization] error.
+     */
+    @objc public static func networkError(withUnderlyingError underlyingError: Error) -> Error {
```
Comment (vegaro) [inline PurchasesCoreSwift/Public/Error Handling/ErrorUtils.swift:23]: I am not sure if these should return `ErrorCode` instead, to indicate that's one of those or `Error` is fine.

File: PurchasesTests/Networking/BackendTests.swift [hunk 6]
```diff
@@ -1502,7 +1502,7 @@ class BackendTests: XCTestCase {
 
         let currentAppUserID = "old id"
         let underlyingErrorMessage = "header fields too large"
-        let underlyingErrorCode = "123456"
+        let underlyingErrorCode = BackendErrorCode.cannotTransferPurchase.rawValue
```
Comment (vegaro) [inline PurchasesTests/Networking/BackendTests.swift:1505]: "123456" would be converted to `.unknownError` with the change from line https://github.com/RevenueCat/purchases-ios/pull/670/files#diff-c4fc22c0d43f6801ea744bd065af825839de0cbe05fdc979640850bcee2135afR278

File: PublicSDKAPITester/RCPurchasesErrorUtilsAPI.m [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+//  RCPurchasesErrorUtilsAPI.m
+//
+//  Created by César de la Vega on 7/21/21.
+
+@import Purchases;
+@import PurchasesCoreSwift;
+
+#import "RCPurchasesErrorUtilsAPI.h"
+
+@implementation RCPurchasesErrorUtilsAPI
```
Comment (aboedo) [inline PublicSDKAPITester/RCPurchasesErrorUtilsAPI.m:19]: why did we ever have this public? it doesn't seem particularly useful to devs

**Review by aboedo — APPROVED** [review]

File: PublicSDKAPITester/RCPurchasesErrorUtilsAPI.m (hunk 7 above)
Comment (vegaro) [reply PublicSDKAPITester/RCPurchasesErrorUtilsAPI.m:19]: Yeah idk. Maybe for testing purposes? So they can create errors like ours. But I think it was just an oversight

**Pull request merged**